src/backend/xdg_desktop_portal.rs

```diff
@@ -0,0 +1,274 @@
+use std::path::PathBuf;
+
+use crate::backend::DialogFutureType;
+use crate::file_dialog::Filter;
+use crate::{FileDialog, FileHandle};
+
+use ashpd::desktop::file_chooser::{
+    FileChooserProxy, FileFilter, OpenFileOptions, SaveFileOptions,
+};
+// TODO: convert raw_window_handle::RawWindowHandle to ashpd::WindowIdentifier
+use ashpd::{zbus, WindowIdentifier};
+
+use smol::block_on;
+
+//
+// Utility functions
+//
+
+fn add_filters_to_open_file_options(
+    filters: Vec<Filter>,
+    mut options: OpenFileOptions,
+) -> OpenFileOptions {
+    for filter in &filters {
+        let mut ashpd_filter = FileFilter::new(&filter.name);
+        for file_extension in &filter.extensions {
+            ashpd_filter = ashpd_filter.glob(&format!("*.{}", file_extension));
+        }
+        options = options.add_filter(ashpd_filter);
+    }
+    options
+}
+
+fn add_filters_to_save_file_options(
+    filters: Vec<Filter>,
+    mut options: SaveFileOptions,
+) -> SaveFileOptions {
+    for filter in &filters {
+        let mut ashpd_filter = FileFilter::new(&filter.name);
+        for file_extension in &filter.extensions {
+            ashpd_filter = ashpd_filter.glob(&format!("*.{}", file_extension));
+        }
+        options = options.add_filter(ashpd_filter);
+    }
+    options
+}
+
+// refer to https://github.com/flatpak/xdg-desktop-portal/issues/213
+fn uri_to_path(uri: &str) -> Option<PathBuf> {
+    uri.strip_prefix("file://").map(PathBuf::from)
+}
+
+//
+// File Picker
+//
+
+use crate::backend::FilePickerDialogImpl;
+impl FilePickerDialogImpl for FileDialog {
+    fn pick_file(self) -> Option<PathBuf> {
+        let connection = block_on(zbus::Connection::session()).unwrap();
```

Comment:
Should the zbus::Connection be saved persistently in the FileDialog struct instead of creating a new one with each function invocation?

Reply:
@bilelmoussaoui what do you think about this?

Reply:
IIUC this would require turning FileDialog and AsyncFileDialog into traits rather than structs so that backends could implement them on different structs.

Reply:
How often you would invoke the function? connecting to the session bus shouldn't be expensive; unless you do multiple calls to various portals at once, there shouldn't be a huge need to store the connection I believe.

Reply:
This is invoked any time the application shows a file dialog. If this is cheap, then I think it's probably not worth the trouble refactoring. @PolyMeilex what do you think?